jupyterhub/auth.py

```diff
@@ -453,7 +453,10 @@ def normalize_username(self, username):
     def check_allowed(self, username, authentication=None):
         """Check if a username is allowed to authenticate based on configuration
 
-        Return True if username is allowed, False otherwise.
+        Return True if username is allowed, False otherwise. Subclasses can also
+        raise a `web.HTTPError(403, message)` to immediately *deny* access and show
+        the end user the given message.
```

Comment:
I definitely agree that documenting the 403 error to enable propagating error messages is a good thing.

But perhaps it is better to point to `check_blocked_users` for _blocking_ users. Since `allowed` may have multiple passes (as oauthenticator's implementation does), putting a rejection in one makes the outcome order-dependent, and may not have the desired effect if another earlier pass granted access. This isn't the case in the base JupyterHub API, which has only the one call. Putting a rejection in `check_blocked_users` will always have the desired effect.

Reply:
In oauthenticator, we are also currently raising an error during the authentication step before check_ functions are called, when a user arrives from an idp that isn't preconfigured via allowed_idps (naming of this is not suitable, its now allowing users from these idps, only accepting users from pre-configured idps)

Reply:
Instead of documenting this here or in other selected methods could we move this to a higher level, and say all methods called during `authenticate()` may raise 403?

Reply:
Yeah, that should be generally true in authenticate, check_allowed/check_blocked_users.